R-package/tests/testthat/test_package.R

```diff
@@ -0,0 +1,352 @@
+#   on CRAN is not upgraded and the older version includes a bug ('TypeError : could not interpret data type') ]
+# reference : https://github.com/scipy/scipy/issues/5353
+
+if (Sys.info()["sysname"] != 'Darwin') {
+
+  # conversion of an R 'dgCMatrix' to a scipy sparse matrix
+  #--------------------------------------------------------
+  
+  testthat::test_that("the 'TO_scipy_sparse' returns an error in case that the input object is not of type 'dgCMatrix' or 'dgRMatrix'", {
+  
+    skip_test_if_no_module("scipy")
+  
+    mt = matrix(runif(20), nrow = 5, ncol = 4)
+  
+    testthat::expect_error( TO_scipy_sparse(mt) )
+  })
+  
+  
+  testthat::test_that("the 'TO_scipy_sparse' returns the correct output", {
+  
+    skip_test_if_no_module("scipy")
+  
+    data = c(1, 0, 2, 0, 0, 3, 4, 5, 6)
+  
+    dgcM = Matrix::Matrix(data = data, nrow = 3,
+  
+                          ncol = 3, byrow = TRUE,
+  
+                          sparse = TRUE)
+  
+    res = TO_scipy_sparse(dgcM)
+  
+    cl_obj = class(res)[1]                                                             # class is python object
+  
+    validate_dims = sum(dim(dgcM) == unlist(reticulate::py_to_r(res$shape))) == 2      # sparse matrix has same dimensions as input R sparse matrix
+  
+    testthat::expect_true( validate_dims && cl_obj == "scipy.sparse.csc.csc_matrix" )
+  })
+  
+  
+  # test that one of the RGF classes works with sparse (scipy) matrices
+  #--------------------------------------------------------------------
+  
+  testthat::test_that("the RGF_Regressor works with sparse (scipy) matrices", {
+  
+    skip_test_if_no_module(c("rgf.sklearn", 'scipy'))
+  
+    set.seed(1)
+    sap = sapply(1:1000, function(x) sample(c(0.0, runif(1)), 1, replace = F))            # create sparse data
+  
+    dgcM = Matrix::Matrix(data = sap,
+  
+                          nrow = 100, ncol = 10,
+  
+                          byrow = TRUE, sparse = TRUE)
+  
+    scipySprse = TO_scipy_sparse(dgcM)                                            # use scipy sparse matrix
+  
+    init_regr = RGF_Regressor$new(max_leaf = 50, sl2 = 0.1, n_iter = 10)
+  
+    init_regr$fit(x = scipySprse, y = y_reg, sample_weight = W)                           # include also a vector of weights
+  
+    pr = init_regr$predict(scipySprse)
+  
+    params = unlist(init_regr$get_params(deep = TRUE))
+  
+    validate = names(params) %in% c("normalize", "loss", "verbose", "algorithm", "n_iter", "learning_rate",
+                                    "sl2", "min_samples_leaf", "opt_interval", "l2", "n_tree_search",
+                                    "reg_depth", "memory_policy", "test_interval", "max_leaf")
+  
+    tmp_score = init_regr$score(x = scipySprse, y = y_reg)
+  
+    tmp_score_W = init_regr$score(x = scipySprse, y = y_reg, sample_weight = W)
+  
+    testthat::expect_true( length(pr) == length(y_reg) && sum(validate) == 15 && is.double(tmp_score) && is.double(tmp_score_W) )
+  })
+}
```

Comment:
We need irregular arguments test. ex. n_estimators < 0.
It is not necessary to examine all the arguments, but it is better for one argument per estimator class to have such a test.

Reply:
You are right. However, since the R-package ports the Python-package I don't think it's wise to repeat the [same tests](https://github.com/RGF-team/rgf/blob/master/python-package/tests/test_rgf_python.py).

Reply:
OK, you don't have to test rgf_python here.
But what happen in R layer if rgf_python raised exception?

`reticulate` defines behavior in such case?
If so, we don't have to test `reticulate` function.

Reply:
I think an example can solve the issue here,

```R
library(RGF)
set.seed(1)
x = matrix(runif(1000), nrow = 100, ncol = 10)
y = runif(100)
RGF_regr = RGF_Regressor$new(max_leaf = -1)
RGF_regr$fit(x, y)

```
<br>
This is the error that I receive using the Rstudio IDE,

```R
 Show Traceback
 
 Rerun with Debug
 Error in py_call_impl(callable, dots$args, dots$keywords) : 
  ValueError: max_leaf must be greater than 0 but was -1. 

```
<br>
Concerning the tests of the *reticulate* package, if you don't mind I would like to keep the tests intact, because I have also to submit the package to CRAN before I update the version for the RGF-team repository and this might lead to issues.